### PR TITLE
Resolve workers being forcelly deleted with `persistent_workers=True`

### LINKED ISSUE #9390: Ability to release resources associated with a DataLoader
https://github.com/pytorch/pytorch/pull/35795 adds `DataLoader.persistent_workers` which is great from a performance point of view. Unfortunately when running multiple trials using Optuna, DataLoader workers take a long time to shut down after a trial completes.

When running under Windows, I am seeing each worker process allocate a commit size of 5GB. When `num_workers = 4` each trial commits 20GB of memory and it takes idle workers a long time (>1 minute) to shut down. If trials are short, I can easily end up with over 100GB of memory committed by worker processes which results in an out of memory error.

Is it possible to proactively shut down DataLoader workers at the end of a trial?

Does `DataLoader` provide a mechanism to release its resources? Does `teardown()` or other hooks provide a mechanism to close data loaders?

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9390

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/loops/test_loops.py [hunk 1]
```diff
@@ -912,7 +912,7 @@ def val_dataloader(self):
 
 
 @RunIf(min_torch="1.8.0")
-@pytest.mark.parametrize("persistent_workers", (True, False))
+@pytest.mark.parametrize("persistent_workers", [False, True])
```
Comment (kaushikb11) [inline tests/loops/test_loops.py:915]: Unnecessary change

**Review by justusschock — APPROVED** [review]

File: tests/loops/test_loops.py [hunk 2]
```diff
@@ -923,7 +923,8 @@ def __init__(self, *args, dataloader: DataLoader, **kwargs):
             self.dataloader = dataloader
 
         def _shutdown_workers(self):
-            setattr(self.dataloader, "has_shutdown_workers", True)
+            has_shutdown_workers = getattr(self.dataloader, "has_shutdown_workers", 0)
```
Comment (kaushikb11) [inline tests/loops/test_loops.py:926]: Need to change the variable from `has_shutdown_workers` to `workers_shutdown`, as the type is changed from Boolean

**Review by rohitgr7 — APPROVED** [review]

File: tests/loops/test_loops.py (hunk 2 above)
Comment (tchaton) [reply tests/loops/test_loops.py:926]: Done !

**Review by kaushikb11 — APPROVED** [review]

**Comment by kaushikb11** [comment]
Let's be careful about the first commit message for the PR. You could update it

**Review by carmocca — APPROVED** [review]
